jetty-websocket/websocket-util-server/src/main/java/org/eclipse/jetty/websocket/util/server/WebSocketUpgradeFilter.java

```diff
@@ -105,22 +109,27 @@ private static FilterHolder getFilter(ServletContext servletContext)
      */
     public static FilterHolder ensureFilter(ServletContext servletContext)
     {
-        FilterHolder existingFilter = WebSocketUpgradeFilter.getFilter(servletContext);
-        if (existingFilter != null)
-            return existingFilter;
-
-        final String name = "WebSocketUpgradeFilter";
-        final String pathSpec = "/*";
-        FilterHolder holder = new FilterHolder(new WebSocketUpgradeFilter());
-        holder.setName(name);
-        holder.setInitParameter(MAPPING_ATTRIBUTE_INIT_PARAM, WebSocketMapping.DEFAULT_KEY);
-
-        holder.setAsyncSupported(true);
-        ServletHandler servletHandler = ContextHandler.getContextHandler(servletContext).getChildHandlerByClass(ServletHandler.class);
-        servletHandler.addFilterWithMapping(holder, pathSpec, EnumSet.of(DispatcherType.REQUEST));
-        if (LOG.isDebugEnabled())
-            LOG.debug("Adding {} mapped to {} in {}", holder, pathSpec, servletContext);
-        return holder;
+        // Lock in case two concurrent requests are initializing the filter lazily.
+        try (AutoLock l = LOCK.lock())
```

Comment:
Are there other `ensureFoo` methods in the websocket setup that also need such locks?

Reply:
I'll approve this PR, but feel free to add other such locks before merging.

Reply:
The other `ensureFoo` methods are all called directly from within the SCI, so we don't get the same sort of race occurring. 
Do you think we should just add the locks there as well anyway?

Reply:
Ok. No need then.